common/set_benchmark.cpp

```diff
@@ -0,0 +1,366 @@
+      // blocking the loop from being optimized away and avoiding disruption of
+      // the generated code that we're benchmarking.
+      benchmark::DoNotOptimize(i);
+
+      bool result = s.BenchContains(lookup_keys[i]);
+      CARBON_DCHECK(result);
+      // We use the lookup success to step through keys, establishing a
+      // dependency between each lookup and allowing us to measure latency
+      // rather than throughput.
+      i += static_cast<ssize_t>(result);
+    }
+  }
+}
+MAP_BENCHMARK_ONE_OP(BM_SetContainsHitPtr, HitArgs);
+
+// Benchmark the "latency" (but more likely the reciprocal throughput, see
+// comment above) of testing for a key in the set that is *not* present.
+template <typename SetT>
+static void BM_SetContainsMissPtr(benchmark::State& state) {
+  using SetWrapperT = SetWrapper<SetT>;
+  using KT = typename SetWrapperT::KeyT;
+  SetWrapperT s;
+  auto [keys, lookup_keys] = GetKeysAndMissKeys<KT>(state.range(0));
+  for (auto k : keys) {
+    s.BenchInsert(k);
+  }
+  ssize_t lookup_keys_size = lookup_keys.size();
+
+  while (state.KeepRunningBatch(lookup_keys_size)) {
+    for (ssize_t i = 0; i < lookup_keys_size;) {
+      benchmark::DoNotOptimize(i);
+
+      bool result = s.BenchContains(lookup_keys[i]);
+      CARBON_DCHECK(!result);
+      i += static_cast<ssize_t>(!result);
+    }
+  }
+}
+MAP_BENCHMARK_ONE_OP(BM_SetContainsMissPtr, SizeArgs);
+
+// A somewhat contrived latency test for the lookup code path.
+//
+// While lookups into a set are often (but not always) simply used to influence
+// control flow, that style of access produces difficult to evaluate benchmark
+// results (see the comments on the `Contains` benchmarks above).
+//
+// So here we actually access the key in the set and convert that key's value to
+// a boolean on the critical path of each iteration. This lets us have a genuine
+// latency benchmark of looking up a key in the set, at the expense of being
+// somewhat contrived. That said, for usage where the key object is queried or
+// operated on in some way once looked up in the set, this will be fairly
+// representative of the latency cost from the data structure.
+template <typename SetT>
+static void BM_SetLookupHitPtr(benchmark::State& state) {
+  using SetWrapperT = SetWrapper<SetT>;
+  using KT = typename SetWrapperT::KeyT;
+  SetWrapperT s;
+  auto [keys, lookup_keys] =
+      GetKeysAndHitKeys<KT>(state.range(0), state.range(1));
+  for (auto k : keys) {
+    s.BenchInsert(k);
+  }
+  ssize_t lookup_keys_size = lookup_keys.size();
+
+  while (state.KeepRunningBatch(lookup_keys_size)) {
+    for (ssize_t i = 0; i < lookup_keys_size;) {
+      benchmark::DoNotOptimize(i);
+
+      bool result = s.BenchLookup(lookup_keys[i]);
+      CARBON_DCHECK(result);
+      i += static_cast<ssize_t>(result);
+    }
+  }
+}
+MAP_BENCHMARK_ONE_OP(BM_SetLookupHitPtr, HitArgs);
+
+// First erase and then insert a key.
```

Comment:
Pretty unclear on why this benchmark is included (for both the map and set benchmarks). This doesn't seem like a common access pattern.